File: openHAB/SwitchUITableViewCell.m
//
//  SwitchUITableViewCell.m
//  openHAB
//
//  Created by Victor Belov on 16/01/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "SwitchUITableViewCell.h"
#import "OpenHABWidget.h"
#import "OpenHABItem.h"

@implementation SwitchUITableViewCell
@synthesize widgetSwitch;

- (id)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        self.widgetSwitch = (UISwitch *)[self viewWithTag:200];
        self.selectionStyle = UITableViewCellSelectionStyleNone;
        self.separatorInset = UIEdgeInsetsZero;
    }
    return self;
}

- (void)displayWidget
{
    self.textLabel.text = [self.widget labelText];
    NSString *state = [self.widget state];
    //if state is nil or empty using the item state ( OH 1.x compatability )
    if(![state length]){
        state = self.widget.item.state;
    }
    if ([self.widget labelValue] != nil)
        self.detailTextLabel.text = [self.widget labelValue];
    else
        self.detailTextLabel.text = nil;
    if ([state isEqualToString:@"ON"]) {
        [self.widgetSwitch setOn:YES];
    } else {
        [self.widgetSwitch setOn:NO];
    }
//    NSLog(@"%f %f %f %f", self.textLabel.frame.origin.x, self.textLabel.frame.origin.y, self.textLabel.frame.size.width, self.textLabel.frame.size.height);
    [self.widgetSwitch addTarget:self action:@selector(switchChange:) forControlEvents:UIControlEventValueChanged];
}

- (void)switchChange:(id)sender{
    if (self.widgetSwitch.isOn) {
        NSLog(@"Switch to ON");
        [self.widget sendCommand:@"ON"];
    } else {
        NSLog(@"Switch to OFF");
        [self.widget sendCommand:@"OFF"];
    }
}

@end

File: openHAB/SwitchUITableViewCell.swift
// Copyright (c) 2010-2023 Contributors to the openHAB project
//
// See the NOTICE file(s) distributed with this work for additional
// information.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// http://www.eclipse.org/legal/epl-2.0
//
// SPDX-License-Identifier: EPL-2.0

import OpenHABCore
import os.log
import UIKit

class SwitchUITableViewCell: GenericUITableViewCell {
    @IBOutlet private var widgetSwitch: UISwitch!

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        initialize()
    }

    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        initialize()
    }

    override func initialize() {
        selectionStyle = .none
        separatorInset = .zero
    }

    override func displayWidget() {
        customTextLabel?.text = widget.labelText
        var state = widget.state
        // if state is nil or empty using the item state ( OH 1.x compatability )
        if state.isEmpty {
            state = (widget.item?.state) ?? ""
        }
        customDetailTextLabel?.text = widget.labelValue ?? ""
        widgetSwitch?.isOn = state.parseAsBool()
        widgetSwitch?.addTarget(self, action: .switchChange, for: .valueChanged)
        super.displayWidget()
    }

    @objc
    func switchChange() {
        if (widgetSwitch?.isOn)! {
            os_log("Switch to ON", log: .viewCycle, type: .info)
            widget.sendCommand("ON")
        } else {
            os_log("Switch to OFF", log: .viewCycle, type: .info)
            widget.sendCommand("OFF")
        }
    }
}

private extension Selector {
    static let switchChange = #selector(SwitchUITableViewCell.switchChange)
}
